src/utils/batch-operate/WindowQueue.ts

```diff
@@ -0,0 +1,180 @@
+		runEmitter: EventEmitter,
+		sessionId: string,
+		client: AirtopClient,
+		operation: (input: BatchOperationInput) => Promise<BatchOperationResponse | undefined>,
+		onError?: (error: BatchOperationError) => Promise<void>,
+	) {
+		if (!Number.isInteger(maxWindowsPerSession) || maxWindowsPerSession <= 0) {
+			throw new Error("maxWindowsPerSession must be a positive integer");
+		}
+
+		this.maxWindowsPerSession = maxWindowsPerSession;
+		this.runEmitter = runEmitter;
+		this.sessionId = sessionId;
+		this.client = client;
+		this.operation = operation;
+		this.onError = onError;
+	}
+
+	public addUrlToQueue(url: BatchOperationUrl): void {
+		this.urlQueue.push(url);
+	}
+
+	private handleHaltEvent(): void {
+		this.client.log("Halt event received");
+		this.isHalted = true;
+	}
+
+	async processInBatches(urls: BatchOperationUrl[]): Promise<void> {
+		this.runEmitter.once("halt", this.handleHaltEvent);
+
+		this.urlQueue = [...urls];
+		this.client.log(
+			`Processing batch: ${JSON.stringify(urls)} for session ${this.sessionId}`,
+		);
+
+		while (this.urlQueue.length > 0) {
+			// Wait for any window to complete before starting a new one
+			if (this.activePromises.length >= this.maxWindowsPerSession) {
+				await Promise.race(this.activePromises);
+				continue;
+			}
+
+			const urlData = this.urlQueue.shift(); // Take the next url from the queue
+			if (!urlData) break; // No more urls to process
+
+			// If we have less than the max concurrent operations, start a new one
+			const promise = (async () => {
+				// Do not process any more urls if the processing has been halted
+				if (this.isHalted) {
+					this.client.log(
+						`Processing halted, skipping window creation for ${urlData.url}`,
+					);
+					return;
+				}
+
+				let windowId: string | undefined;
+				let liveViewUrl: string | undefined;
+				try {
+					// Create a new window pointed to the url
+					this.client.log(
+						`Creating window for ${urlData.url} in session ${this.sessionId}`,
+					);
+					const { data, errors, warnings } = await this.client.windows.create(this.sessionId, {
+						url: urlData.url,
+					});
+					windowId = data.windowId;
+
+					if (warnings) {
+						this.client.warn(`Warnings received creating window ${windowId}: ${JSON.stringify(warnings)}`);
+					}
+
+					if (!windowId) {
+						throw new Error(`WindowId not found, errors: ${JSON.stringify(errors)}`);
+					}
+
+					const { data: windowInfo, warnings: windowWarnings } =
+						await this.client.windows.getWindowInfo(this.sessionId, windowId);
+					liveViewUrl = windowInfo.liveViewUrl;
+
+					if (windowWarnings) {
+						this.client.warn(`Warnings received getting window info for ${windowId}: ${JSON.stringify(windowWarnings)}`);
+					}
+
+					// Run the operation on the window
+					const result = await this.operation({
+						windowId,
+						sessionId: this.sessionId,
+						liveViewUrl,
+						operationUrl: urlData,
+					});
+
+					if (result) {
```

Comment:
Per our conversation offline, I modified the logic so that we return the final list of results after all operations are complete.

I'm happy to rework the logic to fire individual events too, but it sounds like we might want to ship what we have currently and address this in a follow-up.